marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -0,0 +1,734 @@
+          (("roleToken: " <> show assetId <> ": ") <>) <$> do
+            let cardanoAssetId = toCardanoAssetId assetId
+            let quantityMinted = selectAsset value cardanoAssetId
+            check (quantityMinted == 1) ("Expected to mint 1 token, found " <> show quantityMinted)
+      _ -> []
+
+mustSpendRoleTokenViolations
+  :: MarloweVersion v -> Chain.UTxOs -> TxConstraints v -> TxBodyContent BuildTx BabbageEra -> [String]
+mustSpendRoleTokenViolations MarloweV1 utxos TxConstraints{..} TxBodyContent{..} = fold
+  [ passThroughUtxo
+  ]
+  where
+    passThroughUtxo = case roleTokenConstraints of
+      SpendRoleTokens roleTokens -> do
+        roleToken <- Set.toList roleTokens
+        (("roleToken: " <> show roleToken <> ": ") <>) <$> do
+          let
+            isMatch (_, Chain.TransactionOutput{assets = Chain.Assets{tokens = Chain.Tokens tokens}}) =
+              Map.member roleToken tokens
+            mUtxo = find (isMatch . Chain.toUTxOTuple) $ Chain.toUTxOsList utxos
+          case mUtxo of
+            Nothing -> ["UTxO not found that contains role token."]
+            Just utxo ->
+              let
+                (txOutRef, transactionOutput) = Chain.toUTxOTuple utxo
+              in
+                fold
+                  [ check
+                      (any ((== txOutRef) . fromCardanoTxIn . fst) txIns)
+                      ("Expected to consume UTxO " <> show txOutRef)
+                  , check
+                      (any ((== transactionOutput) . fromCardanoTxOut BabbageEra) txOuts)
+                      ("Matching output not found for input " <> show transactionOutput)
+                  ]
+      _ -> []
+
+mustPayToAddressViolations
+  :: MarloweVersion v -> TxConstraints v -> TxBodyContent BuildTx BabbageEra -> [String]
+mustPayToAddressViolations MarloweV1 TxConstraints{..} TxBodyContent{..} = do
+  (address, assets) <- Map.toList payToAddresses
+  (("address: " <> show address <> ": ") <>) <$> do
+    let
+      totalToAddress = foldMap extractValue
+        $ filter ((== address) . extractAddress) txOuts
+    check
+      (totalToAddress == assets)
+      ("Address paid the wrong amount. Expected " <> show assets <> " got " <> show totalToAddress)
+
+mustSendMarloweOutputViolations
+  :: MarloweVersion v
+  -> MarloweContext v
+  -> TxConstraints v
+  -> TxBodyContent BuildTx BabbageEra
+  -> [String]
+mustSendMarloweOutputViolations MarloweV1 MarloweContext{..} TxConstraints{..} TxBodyContent{..} =
+  case marloweOutputConstraints of
+    MarloweOutputConstraintsNone -> []
+    MarloweOutput assets datum ->
+      case find ((== marloweAddress) . extractAddress) txOuts of
+        Nothing -> ["No output found to Marlowe address"]
+        Just txOut -> fold
+          [ check (extractValue txOut == assets) "Wrong assets sent to Marlowe Address."
+          , check (extractDatum txOut == Just (toChainDatum MarloweV1 datum)) "Wrong assets sent to Marlowe Address."
```

Comment:
Shouldn't this error message mention the datum, perhaps "wrong datum sent for Marlowe Address"?

Reply:
Copy :spaghetti: mistakes...